### PR TITLE
[FLINK-8986][e2e-test][WIP] Flink end to end test REST API 

### PR DESCRIPTION
**Pull request by walterddr**
## What is the purpose of the change

Adding end to end test for REST APIs for FLIP-6 generated endpoints

## Brief change log

  - Added `flink-rest-api-test` module in `flink-end-to-end-test` with a periodic stream job and a test suite that runs the REST API tests.
  - Added test script to run the REST API test.

## Verifying this change

N/A, this is a test

## Does this pull request potentially affect one of the following parts:

No

### CODE REVIEW

**Comment by walterddr** [comment]
At this moment it skips 8 of the 39 tests (PATCH & POST methods). And there's 2 test failures and I think should be fix in another JIRA. Namely,
`/jobs/:jobid/vertices/:vertexid/subtasks/:subtaskid/attempts/attempt/accumulators` returns a 500
`/jobs/:jobid/vertices/:vertexid/subtasks/:subtaskid/attempts/attempt/` returns a 500

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/PeriodicStreamingJob.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.SourceFunction;
+import org.apache.flink.types.Row;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ */
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/PeriodicStreamingJob.java]: proper javadoc missing

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+			"Cannot fetch task manager status!");
+		return MAPPER.readValue(resp.getContent(), TaskManagersInfo.class);
+	}
+
+	private static void testMonitoringEndpointSpecs(HttpTestClient httpClient, MessageHeaders spec,
+		Map<String, String> parameterMap) {
+		try {
+			HttpMethodWrapper method = spec.getHttpMethod();
+			String path = getRestEndpointPath(spec.getTargetRestEndpointURL(), parameterMap);
+			if (spec.getRequestClass() == EmptyRequestBody.class) {
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: I would filter out unsupported requests before calling `testMonitoringEndpointSpecs`.

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+	 * @param url target REST endpoint pattern from {@link MessageHeaders} specs
+	 * @param parameterMap parameter replacement map
+	 * @return actual REST URL literal
+	 */
+	private static String getRestEndpointPath(String url, Map<String, String> parameterMap)
+		throws UnsupportedOperationException {
+		for (Map.Entry<String, String> e : parameterMap.entrySet()) {
+			if (e.getValue().equals("")) {
+				if (Pattern.compile(e.getKey()).matcher(url).find()) {
+					throw new UnsupportedOperationException("No parameter replacement found for " + e.getKey());
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: an easier way is to check for `:` in the URL. if it exists there's a parameter that wasn't replaced.

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		String vertexId = jobDetailsInfo.getJobVertexInfos().stream()
+			.findFirst().get().getJobVertexID().toString();
+		parameterMap.put(":vertexid", vertexId);
+		parameterMap.put(":checkpointid", "1"); // test first checkpoint
+		parameterMap.put(":subtaskindex", "0"); // test first subtask
+
+		TaskManagersInfo taskManagersInfo = getTaskManagers(httpClient);
+		String taskMgrId = taskManagersInfo.getTaskManagerInfos().stream().findFirst().get().getResourceId().toString();
+		parameterMap.put(":taskmanagerid", taskMgrId);
+		parameterMap.put(":triggerid", "");
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: this should probably be removed.

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+		final HttpTestClient httpClient = new HttpTestClient(host, port);
+
+		// Validate Flink cluster is running
+		JobIdsWithStatusOverview jobOverview = getJobOverview(httpClient);
+
+		// Get necessary parameters for testing GET endpoints
+		Map<String, String> parameterMap = getParameterMaps(httpClient, jobOverview);
+
+		// Get list of endpoints
+		List<MessageHeaders> specs = new E2ETestDispatcherRestEndpoint().getSpecs();
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: raw parameter usage

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+				" Failures: " + testFailureCount + " Skipped: " + testSkipCount);
+		}
+	}
+
+	@SuppressWarnings("ConstantConditions")
+	private static Map<String, String> getParameterMaps(HttpTestClient httpClient,
+		JobIdsWithStatusOverview jobOverview) throws InterruptedException, IOException, TimeoutException {
+		// Get necessary parameters used for all REST API testings.
+		final Map<String, String> parameterMap = new HashMap<>();
+		Preconditions.checkState(jobOverview.getJobsWithStatus().stream()
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: May be better to make this an assertion, effectively we are relying on the REST API functionality here already.

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+		httpClient.sendGetRequest("/jobs", TEST_TIMEOUT);
+		HttpTestClient.SimpleHttpResponse resp = httpClient.getNextResponse();
+		Preconditions.checkState(resp.getStatus().code() == 200,
+			"Cannot fetch Flink cluster status!");
+		return MAPPER.readValue(resp.getContent(), JobIdsWithStatusOverview.class);
+	}
+
+	private static JobDetailsInfo getJobDetailInfo(HttpTestClient httpClient, String jobId)
+		throws TimeoutException, InterruptedException, IOException {
+			httpClient.sendGetRequest("/jobs/" + jobId, TEST_TIMEOUT);
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: indentation is off

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+	private static JobIdsWithStatusOverview getJobOverview(HttpTestClient httpClient)
+		throws TimeoutException, InterruptedException, IOException {
+		httpClient.sendGetRequest("/jobs", TEST_TIMEOUT);
+		HttpTestClient.SimpleHttpResponse resp = httpClient.getNextResponse();
+		Preconditions.checkState(resp.getStatus().code() == 200,
+			"Cannot fetch Flink cluster status!");
+		return MAPPER.readValue(resp.getContent(), JobIdsWithStatusOverview.class);
+	}
+
+	private static JobDetailsInfo getJobDetailInfo(HttpTestClient httpClient, String jobId)
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: these methods can be made less brittle my using the corresponding `MessageHeader` objects.

```
JobDetailsHeaders headers = JobDetailsHeaders.getInstance();
httpClient.sendGetRequest(headers.getTargetRestEndpointURL().replace(JobIDPathParameter.KEY, jobId), TEST_TIMEOUT);
HttpTestClient.SimpleHttpResponse resp = httpClient.getNextResponse();
Preconditions.checkState(resp.getStatus() == headers.getResponseStatusCode(),
	"Cannot fetch job detail information for job " + jobId);
return MAPPER.readValue(resp.getContent(), headers.getResponseClass());
```

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+				// System.out.println("Found matching status for endpoint " + path + " with method  " + method);
+				@SuppressWarnings("unchecked")
+				Object responseObject = MAPPER.readValue(resp.getContent(), spec.getResponseClass());
+				Preconditions.checkNotNull(responseObject);
+				testSuccessCount += 1;
+			} else {
+				throw new UnsupportedOperationException("Cannot handle REST Test for " + path +
+					" with method " + method + ". Non-empty payload is not supported!");
+			}
+		} catch (IOException |
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: can be simplified to
```
} catch (UnsupportedOperationException e) {
	testSkipCount += 1;
	e.printStackTrace();
} catch (Exception e) {
	testFailureCount += 1;
	e.printStackTrace();
}
```

File: flink-end-to-end-tests/test-scripts/test_rest_api.sh [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+echo "Run Rest-Api-Test Program"
+
+start_cluster
+$FLINK_DIR/bin/taskmanager.sh start
+$FLINK_DIR/bin/taskmanager.sh start
+$FLINK_DIR/bin/taskmanager.sh start
+
+# Start periodic streaming job to test against
+$FLINK_DIR/bin/flink run -p 4 $TEST_JAR_JAR -outputPath file://${TEST_DATA_DIR}/out/result &
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_rest_api.sh]: there's no benefit to running this with a higher parallelism

File: flink-end-to-end-tests/test-scripts/test_rest_api.sh [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+source "$(dirname "$0")"/common.sh
+
+TEST_JAR_JAR=$TEST_INFRA_DIR/../../flink-end-to-end-tests/flink-rest-api-test/target/PeriodicStreamingJob.jar
+TEST_PROGRAM_JAR=$TEST_INFRA_DIR/../../flink-end-to-end-tests/flink-rest-api-test/target/RestApiTestSuite.jar
+
+echo "Run Rest-Api-Test Program"
+
+start_cluster
+$FLINK_DIR/bin/taskmanager.sh start
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_rest_api.sh]: what are the extra taskmanagers for?

File: flink-end-to-end-tests/test-scripts/test_rest_api.sh [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+$FLINK_DIR/bin/taskmanager.sh start
+$FLINK_DIR/bin/taskmanager.sh start
+
+# Start periodic streaming job to test against
+$FLINK_DIR/bin/flink run -p 4 $TEST_JAR_JAR -outputPath file://${TEST_DATA_DIR}/out/result &
+
+# Wait for the job to come up
+sleep 5s
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test_rest_api.sh]: we can handle this more efficiently in the java program by polling the REST API

File: flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+	@SuppressWarnings("ConstantConditions")
+	private static Map<String, String> getParameterMaps(HttpTestClient httpClient,
+		JobIdsWithStatusOverview jobOverview) throws InterruptedException, IOException, TimeoutException {
+		// Get necessary parameters used for all REST API testings.
+		final Map<String, String> parameterMap = new HashMap<>();
+		Preconditions.checkState(jobOverview.getJobsWithStatus().stream()
+			.filter(jobIdWithStatus -> jobIdWithStatus.getJobStatus() == JobStatus.RUNNING)
+			.count() >= 1, "Cannot found active running jobs, discontinuing test!");
+		String jobId = jobOverview.getJobsWithStatus().stream()
+			.filter(jobIdWithStatus -> jobIdWithStatus.getJobStatus() == JobStatus.RUNNING)
```
Comment (zentol) [inline flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java]: replace with `firstMatch`

**Comment by tzulitai** [comment]
Hi @walterddr, what is the status of this PR? Would be nice if we can move forward with this PR (and also the CLI e2e test PR that also you opened.)

**Comment by walterddr** [comment]
Hi @tzulitai . I've actually created a new version of the test based on @zentol 's comment on this PR: https://github.com/walterddr/flink/compare/FLINK-8985...walterddr:FLINK-8986-test

But it actually depends on https://github.com/apache/flink/pull/5863 as I reused the periodic stream job for testing. Is it possible to create a PR on top of another currently pending PR?

**Pull request closed without merging**